api/validators.py

```diff
@@ -0,0 +1,152 @@
+
+resolver = jsonschema.RefResolver('file://' + schema_path + '/schemas/', None)
+
+expected_mongo_schemas = set([
+    'acquisition.json',
+    'collection.json',
+    'container.json',
+    'file.json',
+    'group.json',
+    'note.json',
+    'permission.json',
+    'project.json',
+    'session.json',
+    'subject.json',
+    'user.json',
+    'avatars.json',
+    'tag.json'
+])
+expected_input_schemas = set([
+    'acquisition.json',
+    'collection.json',
+    'container.json',
+    'file.json',
+    'group.json',
+    'note.json',
+    'permission.json',
+    'project.json',
+    'session.json',
+    'subject.json',
+    'user.json',
+    'avatars.json',
+    'download.json',
+    'tag.json'
+])
+mongo_schemas = set()
+input_schemas = set()
+# validate and cache schemas at start time
+for schema_file in os.listdir(schema_path + '/schemas/mongo/'):
+    mongo_schemas.add(schema_file)
+    resolver.resolve('mongo/' + schema_file)
+
+assert mongo_schemas == expected_mongo_schemas, '{} is different from {}'.format(mongo_schemas, expected_schemas)
+
+for schema_file in os.listdir(schema_path + '/schemas/input/'):
+    input_schemas.add(schema_file)
+    resolver.resolve('input/' + schema_file)
+
+assert input_schemas == expected_input_schemas, '{} is different from {}'.format(input_schemas, expected_schemas)
+
+def _validate_json(json_data, schema):
+    jsonschema.validate(json_data, schema, resolver=resolver)
+    #jsonschema.Draft4Validator(schema, resolver=resolver).validate(json_data)
+
+def no_op(g, *args):
+    return g
+
+def mongo_from_schema_file(handler, schema_file):
+    if schema_file is None:
+        return no_op
+    schema = resolver.resolve(schema_file)[1]
+    def g(exec_op):
+        def mongo_val(method, **kwargs):
+            payload = kwargs['payload']
+            log.debug(payload)
+            if method == 'PUT' and schema.get('required'):
+                _schema = copy.copy(schema)
+                _schema.pop('required')
+            else:
+                _schema = schema
+            if method in ['POST', 'PUT']:
+                try:
+                    _validate_json(payload, _schema)
+                except jsonschema.ValidationError as e:
+                    handler.abort(500, str(e))
+            return exec_op(method, **kwargs)
+        return mongo_val
+    return g
+
+def payload_from_schema_file(handler, schema_file):
+    if schema_file is None:
+        return no_op
+    schema = resolver.resolve(schema_file)[1]
+    def g(payload, method):
+        if method == 'PUT' and schema.get('required'):
+            _schema = copy.copy(schema)
+            _schema.pop('required')
+        else:
+            _schema = schema
+        if method in ['POST', 'PUT']:
+            try:
+                _validate_json(payload, _schema)
+            except jsonschema.ValidationError as e:
+                handler.abort(400, str(e))
+    return g
+
+def key_check(handler, schema_file):
```

Comment:
for sublists of mongo container there is no automatic key check when creating, updating or deleting an object. We are adding a custom array field to the json schemas ("key_fields").
The uniqueness is checked on the combination of all the array "key_fields".
For an example you can check https://github.com/scitran/api/blob/refactoring/api/schemas/input/permission.json

So this method ensures that:
1. after a POST and PUT request we don't have two items with the same values for the key set
2. a GET will retrieve a single item
3. a DELETE (most importantly) will delete a single item

I will add a comment with the description.